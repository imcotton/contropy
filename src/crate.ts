import {
    json,
    http_get,
    assert_TrimmedNonEmptyString,
} from './utils.ts';





export function gen (pkg: string, ver: string, {

        agent = 'contropy' as string,
        signal = AbortSignal?.timeout(5_000),

} = {}): Promise<Response> {

    const url = `https://crates.io/api/v1/crates/${ pkg }/${ ver }`;

    return http_get(url, agent, signal);

}





export type Result = Readonly<{
    num: string,
    crate: string,
    checksum: string,
}>;





export function crate (

        pkg: string,
        ver: string,

): Promise<Result> {

    return crate_task(() => gen(pkg, ver));

}





export async function crate_task (

        task: () => Promise<Response>,

): Promise<Result> {

    const res = await json(task()) as {
        version: {
            num: string,
            crate: string,
            checksum: string,
        },
    };

    const { version: { num, crate: crate_, checksum } } = res;

    assert_TrimmedNonEmptyString(num);
    assert_TrimmedNonEmptyString(crate_);
    assert_TrimmedNonEmptyString(checksum);

    return { num, crate: crate_, checksum };

};


