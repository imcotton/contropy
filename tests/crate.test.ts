import c, { agent } from './common.ts';

import { crate_task as task, gen } from '#src/crate.ts';





c.describe('crate', function () {

    const pkgs = [

        [ 'rand', '0.8.5' ],
        [ 'libc', '0.2.153' ],
        [ 'hyper', '1.2.0' ],
        [ 'tokio', '1.36.0' ],
        [ 'regex-syntax', '0.8.2' ],

    ] satisfies Array<[ string, string ]>;

    for (const [ pkg, ver ] of pkgs) {

        c.it_on_ci_and(`has pkg ${ pkg }@${ ver }`, async function (t) {

            const [ signal, abort ] = c.use_signal_abort_controller();

            const res = await task(() => gen(pkg, ver, { agent, signal }));

            await c.assertSnapshot(t, res);

            abort();

        });

    }

});

