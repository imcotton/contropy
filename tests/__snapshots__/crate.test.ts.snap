export const snapshot = {};

snapshot[`crate > has pkg rand@0.8.5 1`] = `
{
  checksum: "34af8d1a0e25924bc5b7c43c079c942339d8f0a8b57c39049bef581b46327404",
  crate: "rand",
  num: "0.8.5",
}
`;

snapshot[`crate > has pkg libc@0.2.153 1`] = `
{
  checksum: "9c198f91728a82281a64e1f4f9eeb25d82cb32a5de251c6bd1b5154d63a8e7bd",
  crate: "libc",
  num: "0.2.153",
}
`;

snapshot[`crate > has pkg hyper@1.2.0 1`] = `
{
  checksum: "186548d73ac615b32a73aafe38fb4f56c0d340e110e5a200bcadbaf2e199263a",
  crate: "hyper",
  num: "1.2.0",
}
`;

snapshot[`crate > has pkg tokio@1.36.0 1`] = `
{
  checksum: "61285f6515fa018fb2d1e46eb21223fff441ee8db5d0f1435e8ab4f5cdb80931",
  crate: "tokio",
  num: "1.36.0",
}
`;

snapshot[`crate > has pkg regex-syntax@0.8.2 1`] = `
{
  checksum: "c08c74e62047bb2de4ff487b251e4a92e24f48745648451635cec7d591162d9f",
  crate: "regex-syntax",
  num: "0.8.2",
}
`;
